### PR TITLE
BUG : Series.to_timestamp and Series.to_period raise user-facing AssertionError

### LINKED ISSUE #33327: BUG: User-facing AssertionError in Series.to_timestamp
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample, a copy-pastable example

```python

In [5]: import pandas as pd

In [6]: pd.Series([0]).to_timestamp()
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-6-1a9d08d7daba> in <module>
----> 1 pd.Series([0]).to_timestamp()

~/sandbox/pandas/pandas/core/series.py in to_timestamp(self, freq, how, copy)
   4565             new_values = new_values.copy()
   4566
-> 4567         assert isinstance(self.index, (ABCDatetimeIndex, ABCPeriodIndex))
   4568         new_index = self.index.to_timestamp(freq=freq, how=how)
   4569         return self._constructor(new_values, index=new_index).__finalize__(

AssertionError:
```

#### Problem description

Users shouldn't see AssertionErrors like this, since they may be disabled when the interpreter starts up.

#### Expected Output

We probably want this to be a `TypeError`.

### PR DESCRIPTION
**Pull request by vipulr8**
- [X] closes #33327
- [X] tests added / passed
- [X] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

### CODE REVIEW

File: pandas/core/series.py [hunk 1]
```diff
@@ -4661,7 +4661,8 @@ def to_timestamp(self, freq=None, how="start", copy=True) -> "Series":
         if copy:
             new_values = new_values.copy()
 
-        assert isinstance(self.index, (ABCDatetimeIndex, ABCPeriodIndex))
+        if not isinstance(self.index, (ABCDatetimeIndex, ABCPeriodIndex)):
+            raise TypeError(f"unsupported Type {self.index}")
```
Comment (simonjayhawkins) [inline pandas/core/series.py:4665]: I _think_ should also raise the TypeError for DatetimeIndex

```
>>> import pandas as pd
>>>
>>> pd.__version__
'1.1.0.dev0+1502.g3ed7dff48'
>>>
>>> rng = pd.date_range("1/1/2012", periods=5, freq="M")
>>>
>>> ts = pd.Series(np.random.randn(len(rng)), index=rng)
>>>
>>> ts
2012-01-31    0.384143
2012-02-29   -0.060017
2012-03-31   -1.506513
2012-04-30    0.898198
2012-05-31    1.035634
Freq: M, dtype: float64
>>>
>>> ts.index
DatetimeIndex(['2012-01-31', '2012-02-29', '2012-03-31', '2012-04-30',
               '2012-05-31'],
              dtype='datetime64[ns]', freq='M')
>>>
>>> ts.to_timestamp()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\simon\pandas\pandas\core\series.py", line 4665, in to_timestamp
    new_index = self.index.to_timestamp(freq=freq, how=how)
AttributeError: 'DatetimeIndex' object has no attribute 'to_timestamp'
```

**Comment by simonjayhawkins** [comment]
Thanks @vipulrai91 for the PR. can you add tests that test for the error message.

**Comment by TomAugspurger** [comment]
Also a whatsnew in doc/source/whatsnew/v1.1.0rst

File: pandas/core/series.py (hunk 1 above)
Comment (vipulrai91) [reply pandas/core/series.py:4665]: Thanks @simonjayhawkins , 
I tried figuring out where I could make this change , but what I have understood is 

error occurs at 
```python
new_index = self.index.to_timestamp(freq=freq, how=how)
```
That means it is passing this step 
```python
isinstance(self.index, (ABCDatetimeIndex, ABCPeriodIndex))
```
which implies DatetimeIndex is sub class / type of either ABCDatetimeIndex, ABCPeriodIndex , have not gone in much details. I am trying to explore more.

Can you guide me where  this change would be required if possible.

**Comment by vipulrai91** [comment]
> Also a whatsnew in doc/source/whatsnew/v1.1.0rst

Yes , I  would add it once the above part is fixed.

**Comment by vipulrai91** [comment]
> Thanks @vipulrai91 for the PR. can you add tests that test for the error message.

@simonjayhawkins 
Should the test be like this 
```python
    @pytest.mark.xfail(raises=TypeError)
    def test_type_error_for_unqualified(self):
        result = pd.Series([0]).to_period()
        assert(result)
```

This one passes.
if this is correct then I intend to add it in

pandas/tests/series/test_period.py

Not sure where to add date 'to_timestamp'

**Comment by simonjayhawkins** [comment]
> Should the test be like this

we use `with pytest.raises(..., match=...)` pattern

> pandas/tests/series/test_period.py
 
pandas\tests\series\methods\test_to_period.py and pandas\tests\series\methods\test_to_timestamp.py


> > Also a whatsnew in doc/source/whatsnew/v1.1.0rst
> 
> Yes , I would add it once the above part is fixed.

add in Backwards incompatible API changes section

As this previously raised AttributeError, https://github.com/pandas-dev/pandas/issues/33327#issuecomment-625849949, can add similar to existing release note for swaplevels

```
- :meth:`DataFrame.swaplevels` now raises a  ``TypeError`` if the axis is not a :class:`MultiIndex`.
  Previously an ``AttributeError`` was raised (:issue:`31126`)
```

**Review by jreback — CHANGES_REQUESTED** [review]
can you add a test that hits this

**Comment by vipulrai91** [comment]
> Thanks @vipulrai91 for the PR. can you add tests that test for the error message.



> can you add a test that hits this

@jreback 
 added tests  at  : pandas/tests/series/methods/test_to_period.py and
       pandas/tests/series/methods/test_to_timestamp.py

Please let me know if this is correct , also , I am not sure which parts to add the doc in doc/source/whatsnew/v1.1.0.rst

File: pandas/tests/series/methods/test_to_period.py [hunk 2]
```diff
@@ -45,3 +46,7 @@ def test_to_period(self):
         expected = df.copy()
         expected.columns = exp_idx
         tm.assert_frame_equal(df.to_period(axis=1), expected)
+
+        # invalid type , #34067 test
+        with pytest.raises(TypeError):
```
Comment (simonjayhawkins) [inline pandas/tests/series/methods/test_to_period.py:51]: can you add match= to test the message

File: pandas/tests/series/methods/test_to_period.py [hunk 3]
```diff
@@ -45,3 +46,7 @@ def test_to_period(self):
         expected = df.copy()
         expected.columns = exp_idx
         tm.assert_frame_equal(df.to_period(axis=1), expected)
+
+        # invalid type , #34067 test
```
Comment (simonjayhawkins) [inline pandas/tests/series/methods/test_to_period.py:50]: can you make this a seperate test, called say test_to_period_raises,

File: pandas/tests/series/methods/test_to_period.py [hunk 4]
```diff
@@ -45,3 +46,7 @@ def test_to_period(self):
...
+        with pytest.raises(TypeError):
+            Series([0]).to_period()
```
Comment (simonjayhawkins) [inline pandas/tests/series/methods/test_to_period.py:52]: maybe use the indices fixture (need to skip on DateTimeIndex)

**Comment by simonjayhawkins** [comment]
> I am not sure which parts to add the doc in doc/source/whatsnew/v1.1.0.rst

see https://github.com/pandas-dev/pandas/pull/34067#issuecomment-627342733

File: doc/source/whatsnew/v1.1.0.rst [hunk 5]
```diff
@@ -339,6 +339,10 @@ Backwards incompatible API changes
 - Combining a ``Categorical`` with integer categories and which contains missing values
   with a float dtype column in operations such as :func:`concat` or :meth:`~DataFrame.append`
   will now result in a float column instead of an object dtyped column (:issue:`33607`)
+- :meth:`Series.to_timestamp` now raises a  ``TypeError`` if the axis is not a :class:`PeriodIndex`.
+  Previously an ``AttributeError`` was raised (:issue:`33327`)
```
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.1.0.rst:343]: can you put this on the previous line

File: pandas/tests/series/methods/test_to_period.py [hunk 6]
```diff
@@ -45,3 +46,10 @@ def test_to_period(self):
         expected = df.copy()
         expected.columns = exp_idx
         tm.assert_frame_equal(df.to_period(axis=1), expected)
+
+    def test_to_period_raises(self, indices):
+        # invalid type , #34067 test
+        idx = indices
+        msg = "unsupported Type RangeIndex"
+        with pytest.raises(TypeError, match=msg):
+            Series([idx]).to_period()
```
Comment (simonjayhawkins) [inline pandas/tests/series/methods/test_to_period.py:55]: we want the series index to be parameterised.

```suggestion
    def test_to_period_raises(self, indices):
        # https://github.com/pandas-dev/pandas/issues/33327
        index = indices
        ser = Series(index=index, dtype=object)
        if not isinstance(index, DatetimeIndex):
            msg = f"unsupported Type {type(index).__name__}"
            with pytest.raises(TypeError, match=msg):
                ser.to_period()
```

File: pandas/core/series.py [hunk 7]
```diff
@@ -4684,7 +4684,8 @@ def to_timestamp(self, freq=None, how="start", copy=True) -> "Series":
         if copy:
             new_values = new_values.copy()
 
-        assert isinstance(self.index, PeriodIndex)
+        if not isinstance(self.index, PeriodIndex):
+            raise TypeError(f"unsupported Type {self.index}")
```
Comment (simonjayhawkins) [inline pandas/core/series.py:4688]:
```suggestion
            raise TypeError(f"unsupported Type {type(self.index).__name__}")
```

File: pandas/core/series.py [hunk 8]
```diff
@@ -4711,7 +4712,8 @@ def to_period(self, freq=None, copy=True) -> "Series":
         if copy:
             new_values = new_values.copy()
 
-        assert isinstance(self.index, DatetimeIndex)
+        if not isinstance(self.index, DatetimeIndex):
+            raise TypeError(f"unsupported Type {self.index}")
```
Comment (simonjayhawkins) [inline pandas/core/series.py:4716]: same

File: pandas/tests/series/methods/test_to_timestamp.py [hunk 9]
```diff
@@ -52,3 +55,18 @@ def _get_with_delta(delta, freq="A-DEC"):
...
         assert result.name == "foo"
+
+    def test_to_timestamp_raises_type_error_for_rangeindex(self, indices):
+        # invalid type , #34067 test
+        idx = indices
+        msg = "unsupported Type RangeIndex"
+        with pytest.raises(TypeError, match=msg):
+            Series([idx]).to_timestamp()
+
+    def test_to_timestamp_raises_type_error_for_datetimeindex(self):
```
Comment (simonjayhawkins) [inline pandas/tests/series/methods/test_to_timestamp.py:66]: only need one test here. like above.

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
Thanks @vipulrai91 for working on this. a few more suggestions but getting close.

File: doc/source/whatsnew/v1.1.0.rst (hunk 5 above)
Comment (vipulrai91) [reply doc/source/whatsnew/v1.1.0.rst]: Did you mean to add it as single line.
Modified it as single line.
Also, other changes updated as suggested.

File: doc/source/whatsnew/v1.1.0.rst [hunk 10]
```diff
@@ -339,6 +339,8 @@ Backwards incompatible API changes
 - Combining a ``Categorical`` with integer categories and which contains missing values
   with a float dtype column in operations such as :func:`concat` or :meth:`~DataFrame.append`
   will now result in a float column instead of an object dtyped column (:issue:`33607`)
+- :meth:`Series.to_timestamp` now raises a  ``TypeError`` if the axis is not a :class:`PeriodIndex`.Previously an ``AttributeError`` was raised (:issue:`33327`)
+- :meth:`Series.to_period` now raises a  ``TypeError`` if the axis is not a :class:`DatetimeIndex`.Previously an ``AttributeError`` was raised (:issue:`33327`)
```
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.1.0.rst:343]:
```suggestion
- :meth:`Series.to_timestamp` now raises a ``TypeError`` if the axis is not a :class:`PeriodIndex`. Previously an ``AttributeError`` was raised (:issue:`33327`)
- :meth:`Series.to_period` now raises a ``TypeError`` if the axis is not a :class:`DatetimeIndex`. Previously an ``AttributeError`` was raised (:issue:`33327`)
```

**Review by simonjayhawkins — APPROVED** [review]
Thanks @vipulrai91 lgtm pending green ex. whitespace in whatsnew.

File: doc/source/whatsnew/v1.1.0.rst (hunk 10 above)
Comment (vipulrai91) [reply doc/source/whatsnew/v1.1.0.rst]: Thanks , updated as suggested.

**Review by simonjayhawkins — APPROVED** [review]
Thanks @vipulrai91 lgtm. @jreback

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @vipulrai91 very nice

**Pull request merged**